nixos/modules/services/security/step-ca.nix

```diff
@@ -0,0 +1,68 @@
+{ config, lib, pkgs, ... }:
+let
+  cfg = config.services.step-ca;
+  settingsFormat = (pkgs.formats.json { });
+in
+{
+  options = {
+    services.step-ca = {
+      enable = lib.mkEnableOption "the smallstep certificate authority server";
+      settings = lib.mkOption {
+        type = lib.types.anything;
+        description = "See https://smallstep.com/docs/step-ca/configuration. https://smallstep.com/docs/step-cli/basic-crypto-operations#run-an-offline-x509-certificate-authority may also be useful. The easiest way is probably to use the existing smallstep tools to set up the CA and then import the config using e.g. `builtins.fromJSON`";
+      };
+      intermediatePasswordFile = lib.mkOption {
+        type = lib.types.str;
+        default = null;
+        example = "/run/keys/smallstep-password";
+        description = "Path to the file containing the password for the intermediate certificate private key. WARNING: make sure to use a quoted absolute path or it is going to be copied to the nix store.";
+      };
+    };
+  };
+
+  config = lib.mkIf config.services.step-ca.enable (
+    let
+      configFile = settingsFormat.generate "ca.json" cfg.settings;
+    in
+    {
+      assertions =
+        [{
+          assertion = !lib.isStorePath cfg.intermediatePasswordFile;
+          message = "`services.step-ca.intermediatePasswordFile` points to a file in the nix store. You should use a quoted absolute path to prevent this.";
+        }];
+
+      systemd.packages = [ pkgs.step-ca ];
+
+      # configuration file indirection is needed to support reloading
+      environment.etc."smallstep-ca.json".source = configFile;
```

Comment:
I think you want `systemd.services.step-ca.restartTriggers = [ cfg.settings ];`

https://search.nixos.org/options?channel=20.09&show=systemd.services.%3Cname%3E.restartTriggers&from=0&size=50&sort=relevance&query=restartTriggers

Reply:
`Environment = [ "" "SMALLSTEP_HACK_DETECT_CONFIG_CHANGE=${configFile}" ]; # I don't know if there is a cleaner way but this forces the activation script to reload this unit on configuration updates.` - this is what `restartTriggers` is. So you want `systemd.services.step-ca.restartTriggers = [ configFile ];`. What @midchildan has may also work if `cfg.settings` is implicitly cast to a path in the nix store, but off the top of my head I'm not sure.

Reply:
FWIW `restartTriggers` can contain anything that can be converted to a string, though a store path would have the benefit of being significantly shorter than the full contents of `cfg.settings`.

From the commit that introduced `restartTriggers`:

> The triggers are just arbitrary strings that are included in the unit
under X-Restart-Triggers.  The idea is that if they change between reconfigurations, switch-to-configuration will restart the unit because its store path changed.

https://github.com/NixOS/nixpkgs/commit/b606165bd9d4c8a1f82160f67bd541f064ed1ddc

Reply:
Sorry, yes I should have been more specific and clarified that point. Thanks.